Poor recovery from [_]::method

Given the following code: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=5f20683e51f9be844a0cc50aeb746f5d

```rust
fn main() {
    let option: Option<&[u8]> = Some(b"...");
    let _ = option.map([_]::to_vec);
}
```

The current output is:

- **it thinks `to_vec` is a crate name**
- **it claims that `[_]` isn't meaningful because it isn't left of an assignment**
- **it claims that 2 arguments were passed to `map`**

```console
error: expected one of `)`, `,`, `.`, `?`, or an operator, found `::`
 --> src/main.rs:3:27
  |
3 |     let _ = option.map([_]::to_vec);
  |                           -^
  |                           |
  |                           expected one of `)`, `,`, `.`, `?`, or an operator
  |                           help: missing `,`

error[E0425]: cannot find crate `to_vec` in the list of imported crates
 --> src/main.rs:3:29
  |
3 |     let _ = option.map([_]::to_vec);
  |                             ^^^^^^ not found in the list of imported crates

error: in expressions, `_` can only be used on the left-hand side of an assignment
 --> src/main.rs:3:25
  |
3 |     let _ = option.map([_]::to_vec);
  |                         ^ `_` not allowed here

error[E0061]: this function takes 1 argument but 2 arguments were supplied
   --> src/main.rs:3:20
    |
3   |     let _ = option.map([_]::to_vec);
    |                    ^^^ ----------- supplied 2 arguments
    |                    |
    |                    expected 1 argument
    |
note: associated function defined here
   --> /media/david/coding/rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/option.rs:843:12
    |
843 |     pub fn map<U, F: FnOnce(T) -> U>(self, f: F) -> Option<U> {
    |            ^^^

Some errors have detailed explanations: E0061, E0425.
For more information about an error, try `rustc --explain E0061`.
```

Ideally the output should look like:

```console
error: missing angle brackets in associated item path
 --> src/main.rs:3:24
  |
3 |     let _ = option.map([_]::to_vec);
  |                        ^^^^^^^^^^^ help: try: `<[_]>::to_vec`
```

Indeed this is exactly that you already get if you had put `[u8]::to_vec` instead of `[_]::to_vec`.

```rust
fn main() {
    let option: Option<&[u8]> = Some(b"...");
    let _ = option.map([u8]::to_vec);
}
```

```console
error: missing angle brackets in associated item path
 --> src/main.rs:3:24
  |
3 |     let _ = option.map([u8]::to_vec);
  |                        ^^^^^^^^^^^^ help: try: `<[u8]>::to_vec`
```

--- a/compiler/rustc_ast/src/ast.rs
+++ b/compiler/rustc_ast/src/ast.rs
@@ -1211,6 +1211,8 @@ impl Expr {
                 }
             }
 
+            ExprKind::Underscore => TyKind::Infer,
+
             // This expression doesn't look like a type syntactically.
             _ => return None,
         };

--- a/src/test/ui/parser/issue-89388.rs
+++ b/src/test/ui/parser/issue-89388.rs
@@ -1,0 +1,7 @@
+// Regression test for #89388.
+
+fn main() {
+    let option: Option<&[u8]> = Some(b"...");
+    let _ = option.map([_]::to_vec);
+    //~^ ERROR: missing angle brackets in associated item path
+}

--- a/src/test/ui/parser/issue-89388.stderr
+++ b/src/test/ui/parser/issue-89388.stderr
@@ -1,0 +1,8 @@
+error: missing angle brackets in associated item path
+  --> $DIR/issue-89388.rs:5:24
+   |
+LL |     let _ = option.map([_]::to_vec);
+   |                        ^^^^^^^^^^^ help: try: `<[_]>::to_vec`
+
+error: aborting due to previous error
+

